File: app/src/main/java/com/github/anrimian/musicplayer/ui/common/theme/ThemeController.java
package com.github.anrimian.musicplayer.ui.common.theme;

import static com.github.anrimian.musicplayer.domain.utils.rx.RxUtils.withDefaultValue;
import static com.github.anrimian.musicplayer.ui.utils.AndroidUtils.getColorFromAttr;

import android.app.Activity;
import android.content.Context;
import android.content.SharedPreferences;
import android.content.res.Configuration;
import android.os.Build;

import androidx.annotation.ColorInt;
import androidx.core.content.ContextCompat;

import com.github.anrimian.musicplayer.R;
import com.github.anrimian.musicplayer.data.utils.preferences.SharedPreferencesHelper;
import com.github.anrimian.musicplayer.ui.utils.AndroidUtils;

import io.reactivex.rxjava3.core.Observable;
import io.reactivex.rxjava3.subjects.BehaviorSubject;

public class ThemeController {

    private static final String PREFERENCES_NAME = "theme_preferences";

    private static final String THEME_ID = "theme_id";
    private static final String LATEST_DARK_THEME_ID = "latest_dark_theme_id";
    private static final String AUTO_DARK_THEME = "auto_dark_theme";
    private static final String FOLLOW_SYSTEM_THEME = "follow_system_theme";

    private final BehaviorSubject<AppTheme> themeSubject = BehaviorSubject.create();

    private final Context context;
    private final SharedPreferencesHelper preferences;

    public ThemeController(Context context) {
        this.context = context;
        SharedPreferences sharedPreferences = context.getSharedPreferences(PREFERENCES_NAME, Context.MODE_PRIVATE);
        this.preferences = new SharedPreferencesHelper(sharedPreferences);
    }

    public void applyCurrentTheme(Activity activity) {
        AppTheme appTheme = getCurrentUsedTheme();
        if (!appTheme.equals(themeSubject.getValue())) {
            themeSubject.onNext(appTheme);
        }
        activity.getTheme().applyStyle(appTheme.getThemeResId(), true);
        updateTaskManager(activity);
    }

    public void applyCurrentSlidrTheme(Activity activity) {
        applyCurrentTheme(activity);
        activity.getTheme().applyStyle(R.style.SlidrActivityTheme, true);
    }

    public void setTheme(Activity activity, AppTheme appTheme) {
        AppTheme currentTheme = getCurrentTheme();
        if (appTheme.equals(currentTheme)) {
            return;
        }
        if (currentTheme.isDark()) {
            setLatestDarkTheme(currentTheme);
        }
        preferences.putInt(THEME_ID, appTheme.getId());

        if (!applyThemeRules(appTheme).equals(applyThemeRules(currentTheme))) {
            processThemeChange(activity, appTheme);
        }
    }

    public AppTheme getCurrentTheme() {
        return AppTheme.getTheme(preferences.getInt(THEME_ID, AppTheme.WHITE_PURPLE_TEAL.getId()));
    }

    @ColorInt
    public int getPrimaryThemeColor() {
        return ContextCompat.getColor(context, getCurrentUsedTheme().getBackgroundColorId());
    }

    public Observable<AppTheme> getAppThemeObservable() {
        return withDefaultValue(themeSubject, this::getCurrentUsedTheme);
    }

    public void setAutoDarkModeEnabled(Activity activity, boolean enabled) {
        AppTheme currentTheme = getCurrentUsedTheme();
        preferences.putBoolean(AUTO_DARK_THEME, enabled);
        if (!currentTheme.equals(getCurrentUsedTheme())) {
            processThemeChange(activity, getCurrentTheme());
        }
    }

    public boolean isAutoDarkThemeEnabled() {
        return preferences.getBoolean(AUTO_DARK_THEME, true);
    }

    public void setFollowSystemThemeEnabled(Activity activity, boolean enabled) {
        AppTheme currentTheme = getCurrentUsedTheme();
        preferences.putBoolean(FOLLOW_SYSTEM_THEME, enabled);
        if (!currentTheme.equals(getCurrentUsedTheme())) {
            processThemeChange(activity, getCurrentTheme());
        }
    }

    public boolean isFollowSystemThemeEnabled() {
        return preferences.getBoolean(FOLLOW_SYSTEM_THEME, Build.VERSION.SDK_INT >= Build.VERSION_CODES.S);
    }

    private AppTheme getLatestDarkTheme() {
        return AppTheme.getTheme(preferences.getInt(LATEST_DARK_THEME_ID, AppTheme.DARK.getId()));
    }

    private void setLatestDarkTheme(AppTheme appTheme) {
        preferences.putInt(LATEST_DARK_THEME_ID, appTheme.getId());
    }

    private void processThemeChange(Activity activity, AppTheme appTheme) {
        activity.getTheme().applyStyle(appTheme.getThemeResId(), true);
        updateTaskManager(activity);

        activity.recreate();

        themeSubject.onNext(appTheme);
    }

    private AppTheme getCurrentUsedTheme() {
        AppTheme theme = AppTheme.getTheme(preferences.getInt(THEME_ID, AppTheme.WHITE_PURPLE_TEAL.getId()));
        return applyThemeRules(theme);
    }

    private AppTheme applyThemeRules(AppTheme theme) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S && isFollowSystemThemeEnabled()) {
            if (isAutoDarkThemeEnabled() && isSystemNightModeEnabled()) {
                return AppTheme.getSystemDarkTheme();
            }
            return AppTheme.getSystemWhiteTheme();
        }
        if (isAutoDarkThemeEnabled()
                && !theme.isDark()
                && isSystemNightModeEnabled()) {
            theme = getLatestDarkTheme();
        }
        return theme;
    }

    private boolean isSystemNightModeEnabled() {
        return (context.getResources().getConfiguration().uiMode & Configuration.UI_MODE_NIGHT_MASK)
                == Configuration.UI_MODE_NIGHT_YES;
    }

    private void updateTaskManager(Activity activity) {
        AndroidUtils.updateTaskManager(activity,
                R.string.app_name,
                activity.getApplicationInfo().icon,
                getColorFromAttr(activity, R.attr.colorPrimary));
    }
}

File: app/src/main/java/com/github/anrimian/musicplayer/ui/common/theme/ThemeController.kt
package com.github.anrimian.musicplayer.ui.common.theme

import android.app.Activity
import android.content.Context
import android.content.res.Configuration
import android.os.Build
import androidx.core.content.ContextCompat
import com.github.anrimian.musicplayer.R
import com.github.anrimian.musicplayer.data.utils.preferences.SharedPreferencesHelper
import com.github.anrimian.musicplayer.domain.utils.rx.RxUtils
import com.github.anrimian.musicplayer.ui.utils.AndroidUtils
import io.reactivex.rxjava3.core.Observable
import io.reactivex.rxjava3.subjects.BehaviorSubject

class ThemeController(private val context: Context) {

    private val themeSubject = BehaviorSubject.create<AppTheme>()
    private val roundCoversSubject = BehaviorSubject.create<Boolean>()
    private val preferences = SharedPreferencesHelper(
        context.getSharedPreferences(PREFERENCES_NAME, Context.MODE_PRIVATE)
    )

    fun applyCurrentTheme(activity: Activity) {
        val appTheme = getCurrentUsedTheme()
        if (appTheme != themeSubject.value) {
            themeSubject.onNext(appTheme)
        }
        activity.theme.applyStyle(appTheme.themeResId, true)
        activity.theme.applyStyle(currentShapeStyle(), true)
        updateTaskManager(activity)
    }

    fun applyCurrentSlidrTheme(activity: Activity) {
        applyCurrentTheme(activity)
        activity.theme.applyStyle(R.style.SlidrActivityTheme, true)
    }

    fun setTheme(activity: Activity, appTheme: AppTheme) {
        val currentTheme = getCurrentTheme()
        if (appTheme == currentTheme) {
            return
        }
        if (currentTheme.isDark) {
            setLatestDarkTheme(currentTheme)
        }
        preferences.putInt(THEME_ID, appTheme.id)
        if (applyThemeRules(appTheme) != applyThemeRules(currentTheme)) {
            processThemeChange(activity, appTheme)
        }
    }

    fun setCircleShapeEnabled(activity: Activity, enabled: Boolean) {
        if (enabled == isCircleShapeEnabled()) {
            return
        }
        preferences.putBoolean(CIRCLE_SHAPE, enabled)
        roundCoversSubject.onNext(enabled)
        activity.recreate()
    }

    fun isCircleShapeEnabled() = preferences.getBoolean(CIRCLE_SHAPE, true)

    fun getCurrentTheme() = AppTheme.getTheme(preferences.getInt(THEME_ID, AppTheme.WHITE_PURPLE_TEAL.id))

    fun getPrimaryThemeColor() = ContextCompat.getColor(context, getCurrentUsedTheme().backgroundColorId)

    fun getAppThemeObservable(): Observable<AppTheme> = RxUtils.withDefaultValue(themeSubject) { getCurrentUsedTheme() }

    fun getRoundCoversObservable(): Observable<Boolean> = RxUtils.withDefaultValue(roundCoversSubject) { isCircleShapeEnabled() }

    fun setAutoDarkModeEnabled(activity: Activity, enabled: Boolean) {
        val currentTheme = getCurrentUsedTheme()
        preferences.putBoolean(AUTO_DARK_THEME, enabled)
        if (currentTheme != getCurrentUsedTheme()) {
            processThemeChange(activity, getCurrentTheme())
        }
    }

    fun isAutoDarkThemeEnabled() = preferences.getBoolean(AUTO_DARK_THEME, true)

    fun setFollowSystemThemeEnabled(activity: Activity, enabled: Boolean) {
        val currentTheme = getCurrentUsedTheme()
        preferences.putBoolean(FOLLOW_SYSTEM_THEME, enabled)
        if (currentTheme != getCurrentUsedTheme()) {
            processThemeChange(activity, getCurrentTheme())
        }
    }

    fun isFollowSystemThemeEnabled() = preferences.getBoolean(
        FOLLOW_SYSTEM_THEME,
        Build.VERSION.SDK_INT >= Build.VERSION_CODES.S
    )

    private fun currentShapeStyle() = if (isCircleShapeEnabled()) R.style.RoundedCorners else R.style.RoundedRectangleCorners

    private fun getLatestDarkTheme() = AppTheme.getTheme(preferences.getInt(LATEST_DARK_THEME_ID, AppTheme.DARK.id))

    private fun setLatestDarkTheme(appTheme: AppTheme) {
        preferences.putInt(LATEST_DARK_THEME_ID, appTheme.id)
    }

    private fun processThemeChange(activity: Activity, appTheme: AppTheme) {
        activity.theme.applyStyle(appTheme.themeResId, true)
        updateTaskManager(activity)
        activity.recreate()
        themeSubject.onNext(appTheme)
    }

    private fun getCurrentUsedTheme(): AppTheme {
        val theme = AppTheme.getTheme(preferences.getInt(THEME_ID, AppTheme.WHITE_PURPLE_TEAL.id))
        return applyThemeRules(theme)
    }

    private fun applyThemeRules(theme: AppTheme): AppTheme {
        var themeResult = theme
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S && isFollowSystemThemeEnabled()) {
            return if (isAutoDarkThemeEnabled() && isSystemNightModeEnabled()) {
                AppTheme.getSystemDarkTheme()
            } else AppTheme.getSystemWhiteTheme()
        }
        if (isAutoDarkThemeEnabled()
            && !themeResult.isDark
            && isSystemNightModeEnabled()
        ) {
            themeResult = getLatestDarkTheme()
        }
        return themeResult
    }

    private fun isSystemNightModeEnabled(): Boolean {
        return (context.resources.configuration.uiMode and Configuration.UI_MODE_NIGHT_MASK
                == Configuration.UI_MODE_NIGHT_YES)
    }

    private fun updateTaskManager(activity: Activity) {
        AndroidUtils.updateTaskManager(
            activity,
            R.string.app_name,
            activity.applicationInfo.icon,
            AndroidUtils.getColorFromAttr(activity, R.attr.colorPrimary)
        )
    }

    private companion object {
        const val PREFERENCES_NAME = "theme_preferences"
        const val THEME_ID = "theme_id"
        const val LATEST_DARK_THEME_ID = "latest_dark_theme_id"
        const val AUTO_DARK_THEME = "auto_dark_theme"
        const val FOLLOW_SYSTEM_THEME = "follow_system_theme"
        const val CIRCLE_SHAPE = "circle_shape"
    }
}